### PR TITLE
fix(integ-tests): cannot make two or more identical assertions

### LINKED ISSUE #22043: (integ-tests): Can't make two awsApiCall for the same service and api in a given context
### Describe the bug

While fumbling around with integration tests I received the following error

`Error: There is already a Construct with name 'AwsApiCallRedshiftdescribeClusters' in Stack [DeployAssert]`

I believe this is because I had two calls for the same service and api

### Expected Behavior

I can make two calls to the same service API combo

### Current Behavior

I can't make two calls to the same service API combo

### Reproduction Steps

```ts
import * as integ from '@aws-cdk/integ-tests';
import * as cdk from '@aws-cdk/core';

const app = new cdk.App();
const stack = new cdk.Stack(app, 'aws-cdk-redshift-cluster-database');


const test = new integ.IntegTest(app, 'aws-cdk-redshift-reboot-test', {
  testCases: [stack],
  stackUpdateWorkflow: true,
  diffAssets: true,
});


const describe = test.assertions.awsApiCall('Redshift', 'describeClusters', {
  ClusterIdentifier: cluster.clusterName,
});

const describe2 = test.assertions.awsApiCall('Redshift', 'describeClusters', {
  ClusterIdentifier: cluster.clusterName,
});

```

### Possible Solution

https://github.com/aws/aws-cdk/blob/1f03e8c40a682a3b5aae90560c84017cfe62762e/packages/%40aws-cdk/integ-tests/lib/assertions/private/deploy-assert.ts#L53

Maybe something like this?

```ts
  private static serviceAPICounter = new Map<string, number>();

...

  public awsApiCall(service: string, api: string, parameters?: any): IAwsApiCall {
    let combo = `AwsApiCall${service}${api}`
    const count = (DeployAssert.serviceAPICounter.get(combo) ?? 0) + 1
    DeployAssert.serviceAPICounter.set(combo, count)
    return new AwsApiCall(this.scope, `${combo}${count}`, {
      api,
      service,
      parameters,
    });
  }
```

### Additional Information/Context

_No response_

### CDK CLI Version

2.41.0

### Framework Version

_No response_

### Node.js Version

v16.16.0

### OS

OsX

### Language

Typescript

### Language Version

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by misterjoshua**
This PR resolves errors in the integration test library when users run identical assertions multiple times.

This change ensures that each use of `awsApiCall`, `httpApiCall`, and `invokeFunction` has a unique identifier, even when multiple identical assertions are used in the same context. We introduced a `uniqueAssertionId` function in the `DeployAssert` construct that maintains backward compatibility with the old id rendering scheme, but prevents the use of conflicting construct ids by tracking the usage of ids and differentiating them where there's a conflict.

Closes #22043, #23049

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Review by lpizzinidev — CHANGES_REQUESTED** [review]
Looks good overall, thanks! 💪
Just a typo and a missing test case and it will be good to go for me.

File: packages/@aws-cdk/integ-tests-alpha/test/assertions/deploy-assert.test.ts [hunk 1]
```diff
@@ -145,6 +145,21 @@ describe('DeployAssert', () => {
...
+      // GIVEN
+      const app = new App();
+
+      // WHEN
+      const deplossert = new DeployAssert(app);
+      deplossert.awsApiCall('MyService', 'MyApi');
+      deplossert.awsApiCall('MyService', 'MyApi');
+
+      // THEN
+      const template = Template.fromStack(deplossert.scope);
```
[the remark this answers is not in the document]
Comment (misterjoshua) [reply packages/@aws-cdk/integ-tests-alpha/test/assertions/deploy-assert.test.ts]: Done!

File: packages/@aws-cdk/integ-tests-alpha/test/assertions/deploy-assert.test.ts [hunk 2]
```diff
@@ -203,6 +218,21 @@ describe('DeployAssert', () => {
...
+      // GIVEN
+      const app = new App();
+
+      // WHEN
+      const deplossert = new DeployAssert(app);
+      deplossert.httpApiCall('https://example.com/test');
+      deplossert.httpApiCall('https://example.com/test');
+
+      // THEN
+      const template = Template.fromStack(deplossert.scope);
```
[the remark this answers is not in the document]
Comment (misterjoshua) [reply packages/@aws-cdk/integ-tests-alpha/test/assertions/deploy-assert.test.ts]: Done.

**Comment by misterjoshua** [comment]
@lpizzinidev Thanks for your review! I believe I've made all the requested adjustments.

**Review by lpizzinidev — APPROVED** [review]
Looks good to me, thanks 👍

**Review by SankyRed — DISMISSED** [review]

**Review by SankyRed — APPROVED** [review]

**Pull request merged**